src/target/riscv/riscv.c

```diff
@@ -1614,6 +1614,429 @@ static int riscv_hit_trigger_hit_bit(struct target *target, uint32_t *unique_id)
+	return (uimm_5to3_bits << 3) + (uimm_7to6_bits << 6);
+}
+
+// c.lq rd_p rs1_p c_uimm9lo c_uimm9hi - offset[7:6] offset[5|4|8]
+static inline uint16_t get_imm_clq(riscv_insn_t instruction)
+{
+	uint16_t uimm_7to6_bits = get_field32(instruction, INSN_FIELD_C_UIMM9LO);
+	uint16_t uimm_5to4and8_bits = get_field32(instruction, INSN_FIELD_C_UIMM9HI);
+	uint16_t uimm_5to4_bits = uimm_5to4and8_bits >> 1;
+	uint16_t uimm_8_bit = uimm_5to4and8_bits & 0x1;
+	return (uimm_5to4_bits << 4) + (uimm_7to6_bits << 6) + (uimm_8_bit << 8);
+}
+
+// c.lqsp rd_n0 c_uimm10sphi c_uimm10splo - offset[5] offset[4|9:6]
+static inline uint16_t get_imm_clqsp(riscv_insn_t instruction)
+{
+	uint16_t uimm_4and9to6_bits =
+		get_field32(instruction, INSN_FIELD_C_UIMM10SPLO);
+	uint16_t uimm_4_bit = uimm_4and9to6_bits >> 4;
+	uint16_t uimm_9to6_bits = uimm_4and9to6_bits & 0xf;
+	uint16_t uimm_5_bit = get_field32(instruction, INSN_FIELD_C_UIMM10SPHI);
+	return (uimm_4_bit << 4) + (uimm_5_bit << 5) + (uimm_9to6_bits << 6);
+}
+
+// c.sqsp c_rs2 c_uimm10sp_s - offset[5:4|9:6]
+static inline uint16_t get_imm_csqsp(riscv_insn_t instruction)
+{
+	uint16_t uimm_5to4and9to6_bits =
+		get_field32(instruction, INSN_FIELD_C_UIMM10SP_S);
+	uint16_t uimm_5to4_biits = uimm_5to4and9to6_bits >> 4;
+	uint16_t uimm_9to6_bits = uimm_5to4and9to6_bits & 0xf;
+	return (uimm_5to4_biits << 4) + (uimm_9to6_bits << 6);
+}
+
+static inline uint32_t get_rs1_c(riscv_insn_t instruction)
+{
+	return 8 + get_field32(instruction, INSN_FIELD_C_SREG1);
+}
+
+uint32_t get_opcode(const riscv_insn_t instruction)
+{
+	/* find out which memory address is accessed by the instruction at dpc */
+	/* opcode is first 7 bits of the instruction */
+	uint32_t opcode = instruction & INSN_FIELD_OPCODE;
+	if ((instruction & 0x03) < 0x03) { // opcode size RVC
+		/* RVC MASK_C = 0xe003 for load/store instructions */
+		opcode = instruction & MASK_C_LD;
+	}
+	return opcode;
+}
+
+static inline int get_storeload_membase_regno(struct target *target,
+		const riscv_insn_t instruction, int *regid)
+{
+	uint32_t opcode = get_opcode(instruction);
+	int rs;
+
+	switch (opcode) {
+	case MATCH_LB:
+	case MATCH_FLH & ~INSN_FIELD_FUNCT3:
+	case MATCH_SB:
+	case MATCH_FSH & ~INSN_FIELD_FUNCT3:
+		rs = get_field32(instruction, INSN_FIELD_RS1);
+		break;
+
+	case MATCH_C_LWSP:
+	case MATCH_C_LDSP: // if xlen >= 64 or MATCH_C_FLWSP:
+	case MATCH_C_FLDSP: // or MATCH_C_LQSP if RV128C
+	case MATCH_C_SWSP:
+	case MATCH_C_SDSP: // if xlen >= 64 or MATCH_C_FSWSP:
+	case MATCH_C_FSDSP: // or MATCH_C_SQSP if xlen == 128
+		rs = GDB_REGNO_SP;
+		break;
+
+	case MATCH_C_LW:
+	case MATCH_C_FLW: // or MATCH_C_LD if xlen >= 64
+	case MATCH_C_FLD: // or MATCH_C_LQ if xlen == 128
+	case MATCH_C_SW:
+	case MATCH_C_FSW: // or MATCH_C_SD if xlen >= 64
+	case MATCH_C_FSD: // or MATCH_C_SQ if xlen == 128
+		rs = get_rs1_c(instruction);
+		break;
+
+	default:
+		LOG_TARGET_DEBUG(target, "%x is not a RV32I or \"C\" load or store",
```

Comment:
[nitpick] [here and around]
`riscv_insn_t` is `uint32_t`, so `%x` is not the correct format specifier.
```suggestion
		LOG_TARGET_DEBUG(target, "%" PRIx32 " is not a RV32I or \"C\" load or store",
```

Reply:
Replaced